LB3210.org

```diff
@@ -977,6 +982,208 @@ Route route = path("users", () ->
 - Deploy in the morning, monitor your dashboards
 - However, "infra tooling" or "platform" team can be helpful
 - The same holds for the "DBA" team
+* A selection of REST patterns
+** Resource tags and caching
+*** Resources have versions
+- Servers can include an =ETag=, which specifies which /version/ of a resource is being served
+#+BEGIN_SRC restclient
+GET http://example.com/widgets/15
+
+200 OK
+Content-Type: application/json
+ETag: "524"
+#+END_SRC
+- No guarantees are made about the content of =ETag=, but often APIs will document what it represents, e.g.
+  * A timestamp of some sort
+  * A monotonically-increasing number
+  * A hash of the latest content
+*** Conditionally retrieving a resource
+- If the latest =ETag= we have seen is "524", we can poll for changes
+- The =If-None-Match= header will /only/ execute our request if the =ETag= has changed
+#+BEGIN_SRC restclient
+GET http://example.com/widgets/15
+If-None-Match: 524
+
+304 Not Modified
+#+END_SRC
+- The server will not send any response if the resource is still at this version
+*** Optimistic offline lock
+- The =ETag= is also useful to make sure nobody else has edited a resource that we're writing back
+- The =If-Match= header will /only/ execute our request if the =ETag= matches
+#+BEGIN_SRC restclient
+PUT http://example.com/widgets/15
+If-Match: 12345
+Content-Type: application/json
+
+{ /* ... some content ..* }
+
+412 Precondition Failed
+#+END_SRC
+*** Resources can be modified
+- Servers can include a =Last-Modified= tag, which specifies /when/ a resource was last changed
+- This can be useful in addition to an =ETag= tag
+#+BEGIN_SRC restclient
+GET http://example.com/widgets/15
+
+200 OK
+Content-Type: application/json
+ETag: "524"
+Last-Modified: Wed, 21 Oct 2015 07:28:00 GMT
+#+END_SRC
+- Request header exist that perform checks against the last-modified date, like =ETag=:
+  * =If-Modified-Since= executes the request /only/ if the last-modified is past the given date
+  * =If-Unmodified-Since= executes the request /only/ if the last-modified is at most the given date
+*** Preventing caching
+- For service-to-service REST calls, we generally don't worry about caching
+- For web browsers, we often want to disallow caching of REST responses
+  * Include =Cache-Control: no-cache=
+** Content-type negotiation
+*** Resource representation
+- The same REST URI is allowed to have several representations
+  * XML, JSON or Protobuf
+  * Short or long
+  * Version 1 or version 2
+*** Specifying resource representation
+- The server specifies the representation of a resource
+  * The =Content-Type= resource header
+- This is typically a well-known value
+  * =text/xml=
+  * =application/json=
+  * =application/protobuf=
+- But it doesn't have to be
+  * =application/vnd.example.myresource.v1+json=
```

Comment:
Consider: vnd=vendor. A vendor specific format as opposed to standardized formats.